### ISSUE TITLE
zookeeper needs to be restarted after upgrading instance type

### ISSUE THREAD

**Issue by jazzl0ver** [cloudstax/firecamp#80]
How to reproduce:
1. Create a working zookeeper/kafka cluster (firecamp version 1.0) using t2.large instances
2. Using CloudFormation update stack control, upgrade the instances to m5.large
3. Zookeeper containers are all up, while Kafka can't start up
4. Restart zookeeper service using firecamp-service-cli

Zookeeper logs before step 4 contain a lot of similar messages:
```
2018-10-11 10:09:13,821 [myid:2] - WARN  [QuorumPeer[myid=2]/0:0:0:0:0:0:0:0:2181:Learner@237] - Unexpected exception, tries=0, connecting to zoo-prod-2.firecamp-prod-firecamp.com/172.31.4.202:2888
java.net.ConnectException: Connection refused (Connection refused)
```

**Comment by JuniusLuo** [comment]
This looks like a rolling restart steps for zk. Would be an issue of zk itself. The full logs would help to further analyse the issue.

**Comment by jazzl0ver** [comment]
The cluster has been running for several months w/o issues and I probably don't have the logs of point of start anymore
[zoo.log.gz](https://github.com/cloudstax/firecamp/files/2479751/zoo.log.gz)
. Please, check the logs attached and let me know if you need something else.

**Comment by jazzl0ver** [comment]
Yeah, I'm not sure as well. The main point is that it's got fixed after restarting zookeeper. Probably we need to wait for all zookeeper nodes to start up before starting up kafka?

**Comment by JuniusLuo** [comment]
Are you able to get logs from node 172.31.4.202?

Kafka does rely on ZK. If zk cluster is not working, kafka will not work. We could consider to introduce the dependency between services. While, how to detecting one service is healthy might not be easy. It has to look into the service internal status. For kafka, it is not necessary to do so. Kafka itself will wait till zk is running.

**Comment by jazzl0ver** [comment]
What kind of logs do you need?

**Comment by jazzl0ver** [comment]
@JuniusLuo, please check https://github.com/cloudstax/firecamp/issues/80#issuecomment-429933249 for zookeeper logs

**Comment by JuniusLuo** [comment]
there is only one log file. We need the log file for zk on 172.31.4.202

**Comment by jazzl0ver** [comment]
[zoo1.log.gz](https://github.com/cloudstax/firecamp/files/2519136/zoo1.log.gz)
[zoo2.log.gz](https://github.com/cloudstax/firecamp/files/2519137/zoo2.log.gz)

**Comment by JuniusLuo** [comment]
It looks like 4.202 is zoo2. zoo2 was not started at 10:10:29. The first log in zoo2.log.gz was at 10:11:14. 
2018-10-11 10:11:14,536 [myid:] - INFO  [main:QuorumPeerConfig@136] - Reading configuration from: /etc/zk/zoo.cfg
2018-10-11 10:11:14,560 [myid:] - INFO  [main:QuorumPeer$QuorumServer@184] - Resolved hostname: zoo-prod-2.firecamp-prod-firecamp.com to address: zoo-prod-2.firecamp-prod-firecamp.com/172.31.4.202

Probably when system was not stable, zk instance kept restarting itself. Restarting the zk service helps to bring all instances up around the same time. This looks like a zk issue itself. Probably zk instance should just wait and retry.

**Comment by jazzl0ver** [comment]
Could you please make firecamp-manager aware of this issue, so Kafka will be restarted only after all ZK instances are up and running?

**Comment by JuniusLuo** [comment]
The manager service aims to be a common service. Monitoring the specific service healthy status is too specific to the service. Looks not the best fit to have the manager service to do this.

**Comment by jazzl0ver** [comment]
@JuniusLuo , in general, I agree with you. But this particular service - Kafka - does not work at all without Zookeeper. So, they're tied by design. Thinking that would be a good feature to handle this case in the manager.
Otherwise, we probably need to put that in docs, so other people won't be messed.

**Comment by JuniusLuo** [comment]
This is not an easy task. This is like requiring the full monitoring ability to ZooKeeper. For example, if ZooKeeper fails because of some bug/issue, Kafka will not work as well.

**Comment by jazzl0ver** [comment]
We just need to start Kafka containers after Zookeeper is got up. Do you really think it's hard to implement?

**Comment by jazzl0ver** [comment]
Understand. What do you think of adding this issue somewhere in the wiki? So people might be aware of such things
